Feature/#7 ouath 회원가입 로그인

CodingApe9 opened the pull request:
## 📄 Summary
> #7 Oauth를 활용한 카카오, 네이버, 구글 로그인, 회원가입 기능 


## 🙋🏻 More
> 변경, 추가된 코드가 많으니 확인하시고 코멘트 많이 남겨주세요!! ❤️

CodingApe9 committed:
feat : Oauth 기본 설계 구축

CodingApe9 committed:
Merge remote-tracking branch 'origin/feature/#7-카카오-회원가입-로그인' into feature/#7-카카오-회원가입-로그인

CodingApe9 committed:
refact : 기능 분리 리펙토링

- oauthClientProvider::CheckOurUser: OauthClient를 가져오는 기능과 우리 서비스의 회원인지 확인하는 기능 분리

CodingApe9 committed:
Merge branch 'master' into feature/#7-Ouath-회원가입-로그인

CodingApe9 committed:
refact : 회원가입 코드 리팩토링

- SignUpService: 코드 재사용을 위해 메서드 시그니처 변경

CodingApe9 committed:
test : 회원가입 테스트 코드 리펙토링

CodingApe9 committed:
refact : 회원가입 코드 리팩토링

- 비밀번호 정규식 user 엔티티로 이전
- NumberUtil 이름 변경 및 패키지 상위로 이동
- 추상화된 타입에 의존하게 변경 @Qualifier 사용

CodingApe9 committed:
fix : 로그인시 비밀번호를 암호화하지 않는 버그 수정

- 비밀번호를 암호화하지 않고 디비에 암호화된 비밀번호화 비교해 항상 비밀번호가 일치하지 않는 버그 수정

CodingApe9 committed:
fix : 로그인 기능 버그 수정

- 가입하지 않은 회원의 응답이 정상적으로 오지 않는 버그 수정

CodingApe9 committed:
feat : 네이버 로그인 기능 추가

- OuathSignupService 메일 전송 기능 추가 필요

CodingApe9 committed:
Merge branch 'feature/#7-Ouath-회원가입-로그인' of /Users/baesuehoon/IdeaProjects/kicketing-back with conflicts.

CodingApe9 committed:
Merge remote-tracking branch 'origin/feature/#7-Ouath-회원가입-로그인' into feature/#7-Ouath-회원가입-로그인

# Conflicts:
#	src/main/java/tback/kicketingback/auth/oauth/controller/CallBackController.java

CodingApe9 committed:
fix : 무작위 비밀번호가 생성되지 않는 버그 수정

CodingApe9 committed:
test : 무작위 비밀번호 생성 기능 테스트 작성

CodingApe9 committed:
refact : 예외 핸들어에 로깅 기능 추가

- 예외 모니터링을 위해 로깅 메시지를 출력하도록 수정

CodingApe9 committed:
feat : 네이버 로그인 기능

- 각 도메인 어스 서버에서 액서스 토큰을 가져와서 우리 서비스의 유저인지 확인
- 첫 회원가입이면 임시 비밀번호를 생성 후 이메일 발송

CodingApe9 committed:
style : 서비스명 오타 수정

- OuathSignupService -> OauthSignupService

CodingApe9 committed:
refact : 네이버 로그인 기능 리펙토링

CodingApe9 committed:
feat : 카카오 로그인 기능

CodingApe9 committed:
refact : 컨트롤러명, 하드코딩 상수로 리펙토링

CodingApe9 committed:
feat : 구글 로그인 기능

pwrwpw commented on src/main/java/tback/kicketingback/user/signin/service/SignInService.java:
```diff
@@ -37,8 +39,8 @@ public TokenResponse signInUser(SignInRequest signInRequest) {
 
-		String password = signInRequest.password();
-		user.validatePassword(password);
+		String encodedPassword = encode(signInRequest.password());
```
이 부분 그냥 singInRequest에서 받을때 처리해도 괜찮을 것 같은데 어떻게 생각하세요?

CodingApe9 replied:
비밀번호 암호화를 유저 엔티티 객체를 생성할때 하자는 의견이 맞을까요?
동의합니다! 서비스보단 유저 엔티티의 도메인 검증 로직이 더 좋을 것 같네요.

pwrwpw commented on src/main/java/tback/kicketingback/user/signup/service/DefaultSignUpService.java:
```diff
@@ -1,48 +1,40 @@
 	@Override
-	public void signUp(SignUpRequest signUpRequest) {
-		if (!isPasswordFormat(signUpRequest.password())) {
+	public void signUp(String name, String email, String password) {
```
이 부분 제 생각에는 그냥 signUpRequest를 받아서 안에서 처리를 하게 된다면 나중에 사용자에게 입력을 더 받아야 할 상황이 생기면 serivce 단만 수정하면 된다고 생각했는데 바꾸신 이유가 있을까요?

cottonrose1011 replied:
저도 Request를 분리한 이유가 궁금해요!

CodingApe9 replied:
우선 결론부터 말씀드리면 오어스 회원가입에서 재활용하기 위해 수정했어요.

기존 사용 형태로 봤을때 SignUpRequest는 사용자에게서 컨트롤러가 받는 DTO의 특성을 띈다고 생각했어요.
그래서 해당 DTO 객체를 서비스에 그대로 전달하는 것은 사용처가 일관적이지 않다는 생각이 들었어요.
또 오어스 회원가입에는 회원가입을 위한 정보를 생성하거나 외부 OAuth를 통해 가져오기 때문에 SignUpRequest 형태를 유지하려면 래핑하는 과정을 거쳐야 했어요.
이 부분에서 오버헤드가 발생할 수 있지 않을까 하는 생각이 들었습니다.

이런 이유들로 SignUpRequest가 아닌 String name, String email, String password 형태로 회원가입 서비스에 전달하게 수정했어요.

추가 의견 남겨주시면 감사하겠습니다. 👍👍👍

pwrwpw replied:
해당 DTO 객체를 서비스에 그대로 전달하는 것은 사용처가 일관적이지 않다는 생각이 들었어요.

이 부분이 잘 이해가 안되는데 원래 DTO 개념이 계층간 데이터 전송을 위해 사용하는 것 아닌가요?
그래서 굳이 계층간 데이터 전송을 위해 DTO를 만들었는데 이걸 재사용을 위해 풀어서 쓰는게 좋은 방식일까는 의문이 듭니다

CodingApe9 replied:
계층간 전송이 아닌 사용자에게 정보를 입력받는 역할을 한다고 생각했습니다.
컨트롤러가 입력받은 형태 그대로 서비스에 전달해도 괜찮을까?
컨트롤러, 서비스 단에서 모두 하나의 DTO에 의존하는 것은 변경에 너무 취약한 코드라고 생각이 들었습니다.
Request를 변경하는 경우 해당 클래스를 의존하는 모든 코드를 수정, 확인해야하니까요.

따라서 서비스에서 불필요한 의존성을 제거하는 방향으로 수정했습니다.

pwrwpw replied:
오히려 컨트롤러에서 꺼내는게 더 DTO에 의존하는게 아닐까요?
만약 request 를 그대로 넘겨준다면 controller에서는 수정할 이유가 없어지는데
현재 코드에는 변경이 일어날 경우 controller와 service와 dto 모두 수정이 필요해 보입니다

pwrwpw replied:
예를 들어 회원가입 시 휴대폰 번호를 꼭 받아야 한다는 요구사항이 추가로 생긴다면

현재 코드에서는 먼저 dto에서 수정을 하고 controller에서도 수정을 해주고 service에서도 수정이 일어나야 하지만

request를 그냥 dto로써 넘겨준다면 dto에서 수정후 service에서만 수정해 주면 될 것 같습니다

CodingApe9 replied:
컨트롤러에서 서비스로 받은 정보를 그대로 전달하는 경우의 수정은 그게 더 간편하겠네요.
하지만 컨트롤러에서 받은 정보를 그대로 서비스단으로 전달하지 않는 경우에는 오히려 현재 상태가 더 유리해 보입니다.
컨트롤러단만 수정하면 되니까요.

하지만 회원가입 기능에 큰 요구사항 변경은 없을 것으로 예상되고 또 코드의 복잡도가 높지 않아 간단히 수정할 수 있을 것 같네요.
결국 어떤 방법을 선택해도 문제는 없어보입니다.

pwrwpw commented on src/main/java/tback/kicketingback/auth/oauth/util/PasswordUtil.java:
```diff
@@ -0,0 +1,39 @@
+import tback.kicketingback.user.exception.exceptions.AuthInvalidPasswordException;
+import tback.kicketingback.utils.NumberCodeUtil;
+
+public class PasswordUtil {
```
저희가 사업자를 낼 수 없으니 생각한 방식인데, 좋은거 같아요

CodingApe9 replied:
😎 기발한 꼼수?였습니다 ㅎㅎ
나중에 이 코드는 비밀번호 찾기 기능에서도 재활용이 가능할 것 같네요.

pwrwpw commented on src/main/java/tback/kicketingback/auth/oauth/domain/OauthClients/NaverOauthClient.java:
```diff
@@ -0,0 +1,85 @@
+
+@Component
+@RequiredArgsConstructor
+public class NaverOauthClient implements OauthClient {
```
Google , Naver가 어쨋든 메서드가 거의 비슷한걸로 보이는데 순서를 맞추면 좀더 보기 편할 것 같아요

CodingApe9 replied:
가독성 측면에서 그게 좋아보이네요 사용 순서에 따라 메서드 위치를 수정했습니다!

pwrwpw commented on src/main/java/tback/kicketingback/user/signup/service/DefaultSignUpService.java:
```diff
@@ -1,48 +1,40 @@
 
-		User user = User.of(signUpRequest.email(), encode(signUpRequest.password()), signUpRequest.username());
-		if (userRepository.existsByEmail(signUpRequest.email())) {
+		User user = User.of(email, encode(password), name);
```
이 부분도 암호화해서 넘겨주는게 아니라 User 안에서 처리하는게 어떨까요?

CodingApe9 replied:
암호화 자체를 유저 도메인으로 넘기자는 의미가 맞을까요?
저도 해당 로직은 도메인 검증에 있는 것이 맞다는 생각이 드네요!
주말에 검증 로직 정리때 같이 수정해봐요!

pwrwpw reviewed (commented):
고생하셨어요~!

cottonrose1011 commented on src/main/java/tback/kicketingback/auth/oauth/exception/exceptions/OAuthResourceAccessFailureException.java:
```diff
@@ -0,0 +1,10 @@
+public class OAuthResourceAccessFailureException extends RuntimeException {
+
+	public OAuthResourceAccessFailureException(OauthClient oauthClient, String accessToken) {
+		super("[%s] [%s]:  Oauth Resource 가져오기 실패".formatted(oauthClient.getClass(), accessToken));
```
예외 던질 때 "~니다"로 통일성 있게 하는게 좋아보입니다!

CodingApe9 replied:
이번 주말 회의때 예외 형식을 정해봐요!
그리고 일괄적으로 수정해봅시다!!

cottonrose1011 commented on src/main/java/tback/kicketingback/user/domain/User.java:
```diff
@@ -60,6 +60,11 @@ private User(final Long id, final String email, final String password, final Str
+		// 디비 설정과 별개.
+		this.state = 0;
+		this.createdAt = LocalDateTime.now();
+		this.modifiedAt = LocalDateTime.now();
```
createdAt, modifiedAt은 master branch에서 분리해놔서 없는게 좋을거 같아요.

CodingApe9 replied:
컨플릭트 해결하면서 자연스럽게 사라졌네요 ㅎㅎ

cottonrose1011 commented on src/test/java/tback/kicketingback/global/encode/PasswordEncoderSHA256Test.java:
```diff
@@ -11,7 +11,7 @@ public class PasswordEncoderSHA256Test {
 		// Given
 
-		final String rawPassword = "123413f";
+		final String rawPassword = "1234abc!@";
```
형식에 맞게 바꾼 부분 좋습니당

cottonrose1011 commented on src/main/java/tback/kicketingback/auth/oauth/dto/ResponseGoogleUser.java:
```diff
@@ -0,0 +1,9 @@
+import com.fasterxml.jackson.annotation.JsonProperty;
+
+public record ResponseGoogleUser(
+	@JsonProperty("id") String id,
+	@JsonProperty("email") String email,
+	@JsonProperty("name") String name) {
```
key값과 같은 값이여도 JsonProperty를 사용하는 것이 좋나요??

CodingApe9 replied:
굳이 사용하지 않아도 문제는 없습니다만 명시적으로 포현하는 편이 코드 이해가 더 쉬울 것으로 생각해 사용했습니다.

CodingApe9 committed:
Merge branch 'master' into feature/#7-Ouath-회원가입-로그인

# Conflicts:
#	src/main/java/tback/kicketingback/global/config/InterceptorAuthConfig.java
#	src/main/java/tback/kicketingback/user/domain/User.java
#	src/main/java/tback/kicketingback/user/signup/controller/SignUpController.java
#	src/main/java/tback/kicketingback/user/signup/service/DefaultSignUpService.java

CodingApe9 committed:
fix : 컨플릭트 해결을 위한 멀지에서 발생한 코드 버그 수정

- 중복 어노테이션 제거
- Request 레코드 인스턴스 변수명 변경으로 인한 수정

CodingApe9 committed:
refact : 구체 OauthClient들 코드 리팩토링

- 가독성 향상을 위한 메서드 위치 조정

CodingApe9 committed:
refact : OAuth 컨트롤러 코드 리팩토링

- 로그인 기능이 헤더에 액서스 토큰을 설정하던 기존 방식이 쿠키로 변경됨에 따라 오어스 로그인, 회원가입도 쿠키에 액서스 토큰을 설정하여 응답하도록 수정

CodingApe9 committed:
refact : 로그인 쿠키 생성 방식 리펙토링

- OAuth 로그인 서비스 교체: 오어스 로그인 서비스 -> 기존 로그인 서비스를 재사용하게 리펙토링
- 토큰을 쿠키로 굽는 방법을 ResponseCookie.from() 빌드 패턴으로 변경.

CodingApe9 committed:
refact : 회원가입 서비스 인자 수정

- 래핑된 SignUpRequest을 전달하게 수정
